### PR TITLE
added possibility to enable/disable regularization for bias

### PR DESCRIPTION
**Pull request by TaplierShiru**
Added possibility to enable/disable regularization for bias (for layers which have bias) and little fix some layers

### CODE REVIEW

File: makiflow/layers/trainable_layers.py [hunk 1]
```diff
@@ -851,8 +851,8 @@ def to_dict(self):
 class BatchNormLayer(BatchNormBaseLayer):
     TYPE = 'BatchNormLayer'
 
-    def __init__(self, D, name, decay=0.9, eps=1e-4, use_gamma=True,
-                 use_beta=True, mean=None, var=None, gamma=None, beta=None, track_running_stats=True):
+    def __init__(self, D, name, decay=0.9, eps=1e-4, use_gamma=True, use_beta=True, regularize_gamma=True,
+                 regularize_beta=True, mean=None, var=None, gamma=None, beta=None, track_running_stats=True):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:855]: Set regularization for this parameter turned off by default. It is actually doesn't make sense to regularize these in case of normalization.

File: makiflow/layers/trainable_layers.py [hunk 2]
```diff
@@ -1002,8 +1003,9 @@ class GroupNormLayer(BatchNormBaseLayer):
     TYPE = 'GroupNormLayer'
     G = 'G'
 
-    def __init__(self, D, name, G=32, decay=0.999, eps=1e-3, use_gamma=True,
-                 use_beta=True, mean=None, var=None, gamma=None, beta=None, track_running_stats=True):
+    def __init__(self, D, name, G=32, decay=0.999, eps=1e-3, use_gamma=True, regularize_gamma=True,
+                 regularize_beta=True, use_beta=True, mean=None, var=None, gamma=None,
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1007]: Set regularization for this parameter turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 3]
```diff
@@ -1189,8 +1192,9 @@ def to_dict(self):
 class NormalizationLayer(BatchNormBaseLayer):
     TYPE = 'NormalizationLayer'
 
-    def __init__(self, D, name, decay=0.999, eps=1e-3, use_gamma=True,
-                 use_beta=True, mean=None, var=None, gamma=None, beta=None, track_running_stats=True):
+    def __init__(self, D, name, decay=0.999, eps=1e-3, use_gamma=True, regularize_gamma=True,
+                 regularize_beta=True, use_beta=True, mean=None, var=None, gamma=None,
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1196]: Set regularization for this parameter turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 4]
```diff
@@ -1356,8 +1361,9 @@ def to_dict(self):
 class InstanceNormLayer(BatchNormBaseLayer):
     TYPE = 'InstanceNormLayer'
 
-    def __init__(self, D, name, decay=0.999, eps=1e-3, use_gamma=True,
-                 use_beta=True, mean=None, var=None, gamma=None, beta=None, track_running_stats=True):
+    def __init__(self, D, name, decay=0.999, eps=1e-3, use_gamma=True, regularize_gamma=True,
+                 regularize_beta=True, use_beta=True, mean=None, var=None, gamma=None,
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1365]: Set regularization for this parameter turned off by default.

File: makiflow/base/maki_entities/input_maki_layer.py [hunk 5]
```diff
@@ -60,5 +60,15 @@ def get_params_dict(self):
         """
         return self._named_params_dict
 
+    def get_params_regularize(self):
+        """
+        This code imitates MakiLayer API for the compatibility sake.
+        This data is used for collect params for regularisation.
+        :return:
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/input_maki_layer.py:67]: The doc string must follow the numpy doc style.

File: makiflow/base/maki_entities/input_maki_layer.py [hunk 6]
```diff
@@ -60,5 +60,15 @@ def get_params_dict(self):
         """
         return self._named_params_dict
 
+    def get_params_regularize(self):
+        """
+        This code imitates MakiLayer API for the compatibility sake.
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/input_maki_layer.py:65]: It does not imitate anything, it is actually a new API. The thing is that it does not violates the old one and fully compatible with it.
Delete this comment.

File: makiflow/base/maki_entities/input_maki_layer.py [hunk 7]
```diff
@@ -60,5 +60,15 @@ def get_params_dict(self):
...
+        This data is used for collect params for regularisation.
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/input_maki_layer.py:66]: Add explanation for why this method actually exists. 
Some of the parameters, like bias, are preferred not to be regularized since it can cause underfitting. Thus, it makes sense to track parameters that are being regularized and that are not.

File: makiflow/base/maki_entities/input_maki_layer.py [hunk 8]
```diff
@@ -60,5 +60,15 @@ def get_params_dict(self):
...
+        :return:
+        list
+            List of parameters for regularisation.
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/input_maki_layer.py:69]: List of parameters to regularized.

File: makiflow/base/maki_entities/maki_layer.py [hunk 9]
```diff
@@ -88,6 +89,15 @@ def get_params_dict(self):
         """
         return self._named_params_dict
 
+    def get_params_regularize(self):
+        """
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/maki_layer.py:93]: Refer to the comment above and fix the same issue.

File: makiflow/layers/trainable_layers.py [hunk 10]
```diff
@@ -36,7 +36,7 @@ class ConvLayer(SimpleForwardLayer):
     INIT_TYPE = 'init_type'
 
     def __init__(self, kw, kh, in_f, out_f, name, stride=1, padding='SAME', activation=tf.nn.relu,
-                 kernel_initializer=InitConvKernel.HE, use_bias=True, W=None, b=None):
+                 kernel_initializer=InitConvKernel.HE, use_bias=True, regularize_bias=True, W=None, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:39]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 11]
```diff
@@ -149,7 +155,7 @@ class UpConvLayer(SimpleForwardLayer):
     INIT_TYPE = 'init_type'
 
     def __init__(self, kw, kh, in_f, out_f, name, size=(2, 2), padding='SAME', activation=tf.nn.relu,
-                 kernel_initializer=InitConvKernel.HE, use_bias=True, W=None, b=None):
+                 kernel_initializer=InitConvKernel.HE, use_bias=True, regularize_bias=True, W=None, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:158]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 12]
```diff
@@ -269,7 +282,7 @@ class BiasLayer(SimpleForwardLayer):
     D = 'D'
     TRAINABLE = 'trainable'
 
-    def __init__(self, D, name, trainable=True, b=None):
+    def __init__(self, D, name, trainable=True, regularize_bias=True, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:285]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 13]
```diff
@@ -338,7 +359,8 @@ class DepthWiseConvLayer(SimpleForwardLayer):
     RATE = 'rate'
 
     def __init__(self, kw, kh, in_f, multiplier, name, stride=1, padding='SAME', rate=[1, 1],
-                 kernel_initializer=InitConvKernel.HE, use_bias=True, activation=tf.nn.relu, W=None, b=None):
+                 kernel_initializer=InitConvKernel.HE, use_bias=True, activation=tf.nn.relu,
+                 regularize_bias=True, W=None, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:363]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 14]
```diff
@@ -462,7 +491,7 @@ class SeparableConvLayer(SimpleForwardLayer):
 
     def __init__(self, kw, kh, in_f, out_f, multiplier, name, stride=1, padding='SAME',
                  dw_kernel_initializer=InitConvKernel.XAVIER_GAUSSIAN_INF, pw_kernel_initializer=InitConvKernel.HE,
-                 use_bias=True, activation=tf.nn.relu,
+                 use_bias=True, regularize_bias=True, activation=tf.nn.relu,
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:494]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 15]
```diff
@@ -596,8 +631,8 @@ class DenseLayer(SimpleForwardLayer):
     USE_BIAS = 'use_bias'
     INIT_TYPE = 'init_type'
 
-    def __init__(self, in_d, out_d, name, activation=tf.nn.relu,
-                 mat_initializer=InitDenseMat.HE, use_bias=True, W=None, b=None):
+    def __init__(self, in_d, out_d, name, activation=tf.nn.relu, mat_initializer=InitDenseMat.HE,
+                 use_bias=True, regularize_bias=True, W=None, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:635]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 16]
```diff
@@ -269,7 +282,7 @@ class BiasLayer(SimpleForwardLayer):
     D = 'D'
     TRAINABLE = 'trainable'
 
-    def __init__(self, D, name, trainable=True, b=None):
+    def __init__(self, D, name, trainable=True, regularize_bias=True, b=None):
         """
         BiasLayer adds a bias vector of dimension D to a tensor.
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:287]: Add commentary on why there is `trainable` argument.
If trainable set to False, then the bias is meant to be a constant. It is done this way to prevent the bias from accidential training if the user simply forgot to make this layer untrainable after creating a model.

File: makiflow/layers/trainable_layers.py [hunk 17]
```diff
@@ -693,7 +735,7 @@ class AtrousConvLayer(SimpleForwardLayer):
     INIT_TYPE = 'init_type'
 
     def __init__(self, kw, kh, in_f, out_f, rate, name, padding='SAME', activation=tf.nn.relu,
-                 kernel_initializer=InitConvKernel.HE, use_bias=True, W=None, b=None):
+                 kernel_initializer=InitConvKernel.HE, use_bias=True, regularize_bias=True, W=None, b=None):
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:738]: Set regularization of the bias turned off by default.

File: makiflow/layers/trainable_layers.py [hunk 18]
```diff
@@ -850,10 +899,12 @@ def _init_train_params(self, data):
 
         self.running_mean = tf.Variable(self.running_mean.astype(np.float32), trainable=False, name=self.name_mean)
         self._named_params_dict[self.name_mean] = self.running_mean
+        self._regularize_params += [self.running_mean]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:902]: Delete it. Running mean must not be regularized. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 19]
```diff
@@ -850,10 +899,12 @@ def _init_train_params(self, data):
...
 
         self.running_variance = tf.Variable(self.running_variance.astype(np.float32), trainable=False,
                                             name=self.name_var)
         self._named_params_dict[self.name_var] = self.running_variance
+        self._regularize_params += [self.running_variance]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:907]: Delete it. Running variance must not be regularized. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 20]
```diff
@@ -1017,10 +1070,12 @@ def _init_train_params(self, data):
...
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1078]: Delete it. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 21]
```diff
@@ -1017,10 +1070,12 @@ def _init_train_params(self, data):
 
         self.running_mean = tf.Variable(self.running_mean.astype(np.float32), trainable=False, name=self.name_mean)
         self._named_params_dict[self.name_mean] = self.running_mean
+        self._regularize_params += [self.running_mean]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1073]: Delete it. Running mean must not be regularized. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 22]
```diff
@@ -1196,9 +1253,12 @@ def _init_train_params(self, data):
...
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1256]: Delete it. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 23]
```diff
@@ -1196,9 +1253,12 @@ def _init_train_params(self, data):
...
+
         self.running_variance = tf.Variable(self.running_variance.astype(np.float32), trainable=False,
                                             name=self.name_var)
         self._named_params_dict[self.name_var] = self.running_variance
+        self._regularize_params += [self.running_variance]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1261]: Delete it. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 24]
```diff
@@ -1363,9 +1425,12 @@ def _init_train_params(self, data):
 
         self.running_mean = tf.Variable(self.running_mean.astype(np.float32), trainable=False, name=self.name_mean)
         self._named_params_dict[self.name_mean] = self.running_mean
+        self._regularize_params += [self.running_mean]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1428]: Delete it. Its value is based purely on statistics, not gradient descent.

File: makiflow/layers/trainable_layers.py [hunk 25]
```diff
@@ -1363,9 +1425,12 @@ def _init_train_params(self, data):
...
+
         self.running_variance = tf.Variable(self.running_variance.astype(np.float32), trainable=False,
                                             name=self.name_var)
         self._named_params_dict[self.name_var] = self.running_variance
+        self._regularize_params += [self.running_variance]
```
Comment (oKatanaaa) [inline makiflow/layers/trainable_layers.py:1433]: Delete it. Its value is based purely on statistics, not gradient descent.

File: makiflow/base/maki_entities/input_maki_layer.py [hunk 26]
```diff
@@ -47,26 +47,35 @@ def __init__(self, data_tensor, name):
...
         This data is used for collect params for regularisation.
-        :return:
+        Some of the parameters, like bias, are preferred not to be regularized since it can cause underfitting.
+        Thus, it makes sense to track parameters that are being regularized and that are not.
+
+        Return
+        ----------
         list
-            List of parameters for regularisation.
+            List of parameters to regularized.
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/input_maki_layer.py:78]: List of parameters to BE regularized.

File: makiflow/base/maki_entities/maki_layer.py [hunk 27]
```diff
@@ -77,24 +77,35 @@ def _training_forward(self, x):
...
         This data is used for collect params for regularisation.
-        :return:
+        Some of the parameters, like bias, are preferred not to be regularized since it can cause underfitting.
+        Thus, it makes sense to track parameters that are being regularized and that are not.
+
+        Return
+        ----------
         list
-            List of parameters for regularisation.
+            List of parameters to regularized.
```
Comment (oKatanaaa) [inline makiflow/base/maki_entities/maki_layer.py:108]: List of parameters to BE regularized.
